### DISCUSSION TITLE
Trying to use presidio analyzer in pyspark.

### DISCUSSION THREAD

**Discussion by Ratnakara-Sarma** [microsoft/presidio#669] [Q&A]
Hello,

I've been trying to use the analyzer in pyspark. My idea is to apply presidio analyzer on a spark dataframe that has file paths(50000 ".txt" files from IMDB reviews dataset found in Kaggle) and their contents. I was able to create the dataframe(50000 x 2). It looks like this:
![image](https://user-images.githubusercontent.com/29379863/115030769-16abf680-9ee5-11eb-922c-96b1ac529f4d.png)

Now I used the python function: 
def pii_apply_pandas(file):
    text = file
    response = analyzer.analyze(correlation_id=0,text = text,entities=[],language='en')
    if len(response) == 0:
        return "no"
    else:
        return "yes"
This function is working fine on single text of any length. Then I converted it into a spark udf. As I try to apply this udf on the spark dataframe, it is taking more than 30 mins and the spark jobs does not even run. Please help.

**Comment by omri374** [comment]
There is an issue with serializing the `en_core_web_lg` spaCy model used by presidio, therefore the sample initializes the model on every request which isn't optimal. Worth checking if the new `en_core_web_trf` model behaves better on Spark.

**Reply by omri374** [reply]
Moved to issue #670

**Reply by omri374** [reply]
Fixed in #672

**Reply by RamSinha** [reply]
Hi @omri374 , I was also trying to run presidio with pyspark and had similar issue. Is there any config in presidio so that it uses `en_core_web_trf ` model?  

btw after upgrading presidio it stopped working with below error.

UserWarning: [W095] Model 'en_core_web_lg' (2.3.0) requires spaCy >=2.3.0,<2.4.0 and is incompatible with the current version (3.0.5). This may lead to unexpected results or runtime errors. To resolve this, download a newer compatible model or retrain your custom model with the current spaCy version. For more details and available updates, run: python -m spacy validate

**Reply by omri374** [reply]
Hi @RamSinha, in terms of adjusting the spaCy model, please see this documentation: https://microsoft.github.io/presidio/analyzer/languages/#configuring-the-nlp-engine

Specifically: 
```python
configuration = {
    "nlp_engine_name": "spacy",
    "models": [{"lang_code": "en", "model_name": "en_core_web_trf"}],
}

# Create NLP engine based on configuration
provider = NlpEngineProvider(nlp_configuration=configuration)
nlp_engine = provider.create_engine()

# Pass the created NLP engine and supported_languages to the AnalyzerEngine
analyzer = AnalyzerEngine(nlp_engine=nlp_engine)
```
Note that this would only work if you have the `en_core_web_trf` model installed.

Regarding your issue, I couldn't reproduce this. Are you working on a fresh installation?

**Reply by RamSinha** [reply]
@omri374 thanks for checking into this, looks like some python packages were missed during installation, after re-installation it started to work.. However the problem with pyspark still persist [the job doesn't proceed and after sometime it crashes].

As a workaround i have created k8 deployment with presidio and making http call from my scala/spark code.

**Reply by omri374** [reply]
@RamSinha thanks for letting us know. We'd be happy to get some more information on the Spark issue you were facing. If there are any exceptions / log traces we could look at that would be great.

**Reply by RamSinha** [reply]
So when i try to call analyzer.analyze() method inside my executor code (), job goes into the hung state. 
Below is the code skeleton. 
In this case: The spark goes into hung state, doesn't do anything.. and eventually crashes with out of memory (even when df has only 4 sample records).. 

```
analyzer = AnalyzerEngine()
def checkRecord(rows):
    // apply some transformation
   // Return list of records that may have PII

df.rdd.mapPartitions(checkRecord)
```

Also if i move analyzer initialization under checkRecord method- then i get below error
```
objc[49543]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called.
objc[49543]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

**Reply by omri374** [reply]
Have you looked at our sample implementation here: https://github.com/microsoft/presidio/blob/main/docs/samples/deployments/spark/presidio_anonymize_blobs.py

You seem to use a different implementation and it is difficult to say if this is an issue with Presidio/spaCy or with the spark process you are running.

**Reply by omri374** [reply]
Spark has to broadcast the AnalyzerEngine (together with the underlying spaCy model) to all workers.

**Reply by RamSinha** [reply]
Thanks for the pointer, lemme restructure the code based on above hints.

**Reply by prasanthgp93** [reply]
[omri374](https://github.com/omri374),[RamSinha](https://github.com/RamSinha)

We have a similar issue here to parse on larger dataset you got any optimal solution here.
we are using same pyspark and below are the packages we are downloading below packages.
Also we have use case of records having text length more than 3 lakhs to 9 lakhs ,

Used : Apache Spark 3.2.1,Python version 3.8.10

#!/bin/bash
pip install presidio-analyzer
pip install presidio-anonymizer
python -m spacy download en_core_web_lg

Please do reply
